event_sourcery_sqlalchemy/models.py

```diff
@@ -12,20 +28,56 @@ def configure_models(base: Type) -> None:
         registry(metadata=base.metadata, class_registry={}).map_declaratively(model_cls)
 
 
+class StreamIdComparator(Comparator[StreamId]):
+    def __init__(self, uuid: Column, name: Column, category: Column) -> None:
+        super().__init__(Composite(uuid, name, category))
+
+    def __eq__(self, other: Any) -> ColumnElement[bool]:
+        uuid, name, category = cast(Composite, self.__clause_element__()).attrs
+        same_stream_id = cast(
+            ColumnElement[Type[bool]],
+            name == other.name if other.name else uuid == other,
```

Comment:
Maybe I misunderstood query that I've moved.

```python
events_stmt = select(EventModel).order_by(EventModel.version)
events_stmt = events_stmt.filter(EventModel.stream_id == stream_id)

if stream_id.name is not None:
    stream_id_subq = (
        select(StreamModel.uuid)
        .filter(StreamModel.name == stream_id.name)
        .scalar_subquery()
    )
    events_stmt = events_stmt.filter(EventModel.stream_id == stream_id_subq)
```

I assumed originally it was using UUID to find a stream.

```python
events_stmt = select(EventModel).order_by(EventModel.version)
events_stmt = events_stmt.filter(EventModel.stream_id == stream_id)
```

In case we have a name it overrides the query to use only `name` to find UUID in DB and use these entries.
```python
if stream_id.name is not None:
    stream_id_subq = (
        select(StreamModel.uuid)
        .filter(StreamModel.name == stream_id.name)
        .scalar_subquery()
    )
    events_stmt = events_stmt.filter(EventModel.stream_id == stream_id_subq)
```

Is this right?
So if I've got it right we can discuss changing this behavior, but in separate refactoring. If I missed something then I will fix it.

Reply:
```
events_stmt = events_stmt.filter(EventModel.stream_id == stream_id_subq)
```
As far as I know SQLAlchemy, this statement under `if` won't overwrite filter on the field but will append the second one so in the resultant query there will be
`WHERE stream_id = :uuid and stream_id = (SELECT ... )`

We can talk about it - but I assume the behaviour should be exactly as you described

Reply:
You are right it won't replace, just add an additional subquery. Still missing a test for it.

I will fix it.

Reply:
if that's fixed, we can go forward with this PR